kvserver/closedts: add metrics for policy refresher

wenyihu6 opened the pull request:
**kvserver: add kv.closed_timestamp.policy_change**

Previously, it was difficult to measure how often policies changed for ranges,
which is important because such changes can trigger additional range updates
sent in side transport.

This commit adds a metric to track the number of policy changes on replicas.

Part of: #143890
Release note: none

---

**kvserver: add more metrics for policies**

Previously, it was difficult to determine how many ranges fell into each latency
bucket policy. This commit adds 18 new metrics to StoreMetrics to track the
number of ranges per policy bucket for every store.

Part of: #143890
Release note: none

---

**kvserver: add kv.closed_timestamp.policy_latency_info_missing**

When a replica refreshes its policies, it looks up its peer replicas latency
info via a map passed by PolicyRefresher, which in turn periodically pulls node
latency info from RPCContext. If latency data for a node is missing, a default
hardcoded max RTT of 150ms is used.

Previously, it was hard to tell when this is happening. This commit adds metrics
to track how often the closed timestamp policy refresh falls back to the default
RTT due to missing node latency info. A high count might indicate the latency
cache isn’t refreshed frequently enough, suggesting we should consider lowering
kv.closed_timestamp.policy_latency_refresh_interval.

Resolves: #143890
Release note: none

wenyihu6 committed:
kvserver: add kv.closed_timestamp.policy_change

Previously, it was difficult to measure how often policies changed for ranges,
which is important because such changes can trigger additional range updates
sent in side transport.

This commit adds a metric to track the number of policy changes on replicas.

Part of: #143890
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144518)
<!-- Reviewable:end -->

wenyihu6 committed:
kvserver: add more metrics for policies

Previously, it was difficult to determine how many ranges fell into each latency
bucket policy. This commit adds 18 new metrics to StoreMetrics to track the
number of ranges per policy bucket for every store.

Part of: #143890
Release note: none

wenyihu6 committed:
kvserver: add kv.closed_timestamp.policy_latency_info_missing

When a replica refreshes its policies, it looks up its peer replicas latency
info via a map passed by PolicyRefresher, which in turn periodically pulls node
latency info from RPCContext. If latency data for a node is missing, a default
hardcoded max RTT of 150ms is used.

Previously, it was hard to tell when this is happening. This commit adds metrics
to track how often the closed timestamp policy refresh falls back to the default
RTT due to missing node latency info. A high count might indicate the latency
cache isn’t refreshed frequently enough, suggesting we should consider lowering
kv.closed_timestamp.policy_latency_refresh_interval.

Resolves: #143890
Release note: none

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r1, 4 of 4 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144518#-:-OOZij6yB7H9hrE9U9Vf:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/replica.go` line 1385 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144518#-OOZhxhzBen2nZKGNNUP:-OOZhxhzBen2nZKGNNUQ:b-2qluoj) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0f6d80b78d74f0e012e3033c7d2de0d8ca71924/pkg/kv/kvserver/replica.go#L1385)):*
> ```Go
> 	newPolicy := computeNewPolicy(oldPolicy)
> 	r.cachedClosedTimestampPolicy.Store(int32(newPolicy))
> 	if oldPolicy != newPolicy {
> ```

nit: 
```golang

ctpb.RangeClosedTimestampPolicy(r.cachedClosedTimestampPolicy.Load()) != newPolic {
...
// update metric
// store new policy
}

```


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144518)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani)

___
*[`pkg/kv/kvserver/replica.go` line 1385 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/144518#-OOZhxhzBen2nZKGNNUP:-OO_4OK3AK2hcZMVJIoB:b9m4gee) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0f6d80b78d74f0e012e3033c7d2de0d8ca71924/pkg/kv/kvserver/replica.go#L1385)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

nit: 
```golang

ctpb.RangeClosedTimestampPolicy(r.cachedClosedTimestampPolicy.Load()) != newPolic {
...
// update metric
// store new policy
}

```
</blockquote></details>

Moved the policy update to be inside the if statement. I think this is what you had in mind, but lmk if I'm off. We need to instantiate `oldPolicy` as a separate variable because it's required by `computeNewPolicy` for dampening and also needed for policy comparison metrics update. 


<!-- Sent from Reviewable.io -->

wenyihu6 commented:
TFTR!

bors r=arulajmani
